src/rsa/blinding.rs

```diff
@@ -12,52 +12,127 @@
 // OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 // CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
-use error;
-use super::bigint;
+//! RSA base blinding.
+//!
+//! This is based on the suggestions in the paper
+//! "Timing Attacks on Implementations of Diffie-Hellman, RSA, DSS, and Other
+//! Systems" by Paul C. Kocher.
 
-pub struct Blinding(*mut BN_BLINDING);
+use {error, rand};
+use core;
+use super::{bigint, N};
 
-impl Drop for Blinding {
-    fn drop(&mut self) { unsafe { GFp_BN_BLINDING_free(self.as_mut_ref()) } }
-}
+pub struct Blinding(Option<Contents>);
 
-// `Blinding` uniquely owns and references its contents.
-unsafe impl Send for Blinding {}
+struct Contents {
+    blinding_factor: bigint::Elem<N>, // `(1 / v_i)**e` from the paper.
+    blinding_factor_inv: bigint::Elem<N>, // `v_f` from the paper.
+    remaining: usize,
+}
 
 impl Blinding {
-    pub fn new() -> Result<Blinding, error::Unspecified> {
-        let r = unsafe { GFp_BN_BLINDING_new() };
-        if r.is_null() {
-            return Err(error::Unspecified);
-        }
-        Ok(Blinding(r))
+    pub fn new() -> Self { Blinding(None) }
+
+    pub fn blind<F>(&mut self, x: bigint::ElemDecoded<N>,
```

Comment:
Thanks for the feedback.

The distinction between `RSASigningState` and `RSAKeyPair` is that `RSASigningState` is per-thread and `RSAKeyPair` is the state shared (immutably) across all threads by all the `RSASigningState` instances for the same key. The idea is that a high-performance server would likely create one `RSASigningState` per core (perhaps in a thread-local) per key. So, we generally want to make the `RSASigningState`s small.

Here's the definition of `RSASigningState`:
```rust
pub struct RSASigningState {
    key_pair: std::sync::Arc<RSAKeyPair>,
    blinding: blinding::Blinding,
}
```

We could instead define `RSASigningState` as a newtype:
```rust
pub struct RSASigningState(blinding::Blinding)
```
Then we'd move the `key_pair` field to the blinding, without increasing the size of `RSASigningState`.  After all, the idea of `RSASigningState` is to be the abstraction "a thing you need exclusive access to to sign with an RSA key, for some reason", and `Blinding` is the reason you need some extra thing. That would make it clear that the blinding values are specific to that key pair and that there's no mixing of key pairs for the same blinding. Then `RSASigningState::sign()` would borrow the `key_pair` out of the `Blinding`.

OTOH, it's a bit weird that the `Blinding` should know about the entire contents of the `RSAKeyPair` when only `n` and `e` are relevant to it. And it's also weird that it would need to know about `Arc` and threading behavior. What I'm hoping is that instead we'll be able to expand on the use of the `field: PhantomData<Field>` construct using the "lightweight static capabilities" idea to ensure that one can never mix two different tuples of (`N`, `P`, `Q`, ...). But we haven't implemented the important parts that actually enforce that yet, so maybe it is presumptuous to act like we did (or that we even know how to yet!).

Anyway, I'd love to hear what people think about this.